File: backend/src/main/java/com/memail/service/LabelService.java
package com.memail.service;

import com.memail.model.Label;
import com.memail.model.MessageLabel;
import com.memail.repository.LabelRepository;
import com.memail.repository.MessageLabelRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class LabelService {

    @Autowired
    private LabelRepository labelRepository;

    @Autowired
    private MessageLabelRepository messageLabelRepository;

    /**
     * Get all labels for a user
     */
    public List<Label> getUserLabels(String userId) {
        return labelRepository.findByUserIdOrderByNameAsc(userId);
    }

    /**
     * Create a new label for a user
     */
    public Label createLabel(String userId, String name, String color) {
        if (labelRepository.existsByUserIdAndName(userId, name)) {
            throw new IllegalArgumentException("Label with name '" + name + "' already exists");
        }

        Label label = new Label(userId, name, color);
        return labelRepository.save(label);
    }

    /**
     * Update an existing label
     */
    public Label updateLabel(String userId, Long labelId, String name, String color) {
        Optional<Label> labelOpt = labelRepository.findByUserIdAndId(userId, labelId);
        if (labelOpt.isEmpty()) {
            throw new IllegalArgumentException("Label not found or does not belong to user");
        }

        Label label = labelOpt.get();

        // Check if new name conflicts with existing labels (excluding current label)
        if (!label.getName().equals(name) && labelRepository.existsByUserIdAndName(userId, name)) {
            throw new IllegalArgumentException("Label with name '" + name + "' already exists");
        }

        label.setName(name);
        label.setColor(color);
        return labelRepository.save(label);
    }

    /**
     * Delete a label and all its message associations
     */
    public void deleteLabel(String userId, Long labelId) {
        Optional<Label> labelOpt = labelRepository.findByUserIdAndId(userId, labelId);
        if (labelOpt.isEmpty()) {
            throw new IllegalArgumentException("Label not found or does not belong to user");
        }

        // Delete all message label associations first
        messageLabelRepository.deleteByLabelId(labelId);

        // Then delete the label itself
        labelRepository.deleteById(labelId);
    }

    /**
     * Get a specific label by ID for a user
     */
    public Optional<Label> getLabel(String userId, Long labelId) {
        return labelRepository.findByUserIdAndId(userId, labelId);
    }

    /**
     * Apply a label to a message
     */
    public MessageLabel applyLabelToMessage(String userId, String messageUid, String folder, Long labelId) {
        // Verify the label belongs to the user
        Optional<Label> labelOpt = labelRepository.findByUserIdAndId(userId, labelId);
        if (labelOpt.isEmpty()) {
            throw new IllegalArgumentException("Label not found or does not belong to user");
        }

        // Check if the label is already applied to this message
        if (messageLabelRepository.existsByUserIdAndMessageUidAndFolderAndLabelId(userId, messageUid, folder, labelId)) {
            throw new IllegalArgumentException("Label is already applied to this message");
        }

        Label label = labelOpt.get();
        MessageLabel messageLabel = new MessageLabel(userId, messageUid, folder, label);
        return messageLabelRepository.save(messageLabel);
    }

    /**
     * Remove a label from a message
     */
    public void removeLabelFromMessage(String userId, String messageUid, String folder, Long labelId) {
        // Verify the label belongs to the user
        if (!labelRepository.findByUserIdAndId(userId, labelId).isPresent()) {
            throw new IllegalArgumentException("Label not found or does not belong to user");
        }

        messageLabelRepository.deleteByUserIdAndMessageUidAndFolderAndLabelId(userId, messageUid, folder, labelId);
    }

    /**
     * Get all labels applied to a specific message
     */
    public List<MessageLabel> getMessageLabels(String userId, String messageUid, String folder) {
        return messageLabelRepository.findByUserIdAndMessageUidAndFolder(userId, messageUid, folder);
    }

    /**
     * Get all messages with a specific label
     */
    public List<MessageLabel> getMessagesWithLabel(String userId, Long labelId) {
        // Verify the label belongs to the user
        if (!labelRepository.findByUserIdAndId(userId, labelId).isPresent()) {
            throw new IllegalArgumentException("Label not found or does not belong to user");
        }

        return messageLabelRepository.findByUserIdAndLabelIdOrderByCreatedAtDesc(userId, labelId);
    }

    /**
     * Check if a message has a specific label
     */
    public boolean messageHasLabel(String userId, String messageUid, String folder, Long labelId) {
        return messageLabelRepository.existsByUserIdAndMessageUidAndFolderAndLabelId(userId, messageUid, folder, labelId);
    }

    /**
     * Apply labels to multiple messages in batch
     */
    public void applyLabelsToMessages(String userId, List<String> messageUids, String folder, List<Long> labelIds) {
        // Verify all labels belong to the user
        for (Long labelId : labelIds) {
            if (!labelRepository.findByUserIdAndId(userId, labelId).isPresent()) {
                throw new IllegalArgumentException("Label with ID " + labelId + " not found or does not belong to user");
            }
        }

        for (String messageUid : messageUids) {
            for (Long labelId : labelIds) {
                // Only apply if not already applied
                if (!messageLabelRepository.existsByUserIdAndMessageUidAndFolderAndLabelId(userId, messageUid, folder, labelId)) {
                    Label label = labelRepository.findByUserIdAndId(userId, labelId).get();
                    MessageLabel messageLabel = new MessageLabel(userId, messageUid, folder, label);
                    messageLabelRepository.save(messageLabel);
                }
            }
        }
    }

    /**
     * Remove labels from multiple messages in batch
     */
    public void removeLabelsFromMessages(String userId, List<String> messageUids, String folder, List<Long> labelIds) {
        // Verify all labels belong to the user
        for (Long labelId : labelIds) {
            if (!labelRepository.findByUserIdAndId(userId, labelId).isPresent()) {
                throw new IllegalArgumentException("Label with ID " + labelId + " not found or does not belong to user");
            }
        }

        for (Long labelId : labelIds) {
            for (String messageUid : messageUids) {
                messageLabelRepository.deleteByUserIdAndMessageUidAndFolderAndLabelId(userId, messageUid, folder, labelId);
            }
        }
    }

    /**
     * Get message UIDs for a specific label and folder
     */
    public List<String> getMessageUidsWithLabel(String userId, Long labelId, String folder) {
        // Verify the label belongs to the user
        if (!labelRepository.findByUserIdAndId(userId, labelId).isPresent()) {
            throw new IllegalArgumentException("Label not found or does not belong to user");
        }

        return messageLabelRepository.findMessageUidsByUserIdAndLabelIdAndFolder(userId, labelId, folder);
    }

    /**
     * Get label count for a specific label
     */
    public long getLabelUsageCount(String userId, Long labelId) {
        // Verify the label belongs to the user
        if (!labelRepository.findByUserIdAndId(userId, labelId).isPresent()) {
            throw new IllegalArgumentException("Label not found or does not belong to user");
        }

        return messageLabelRepository.countByUserIdAndLabelId(userId, labelId);
    }

    /**
     * Delete all data for a user (for user deletion)
     */
    public void deleteAllUserData(String userId) {
        messageLabelRepository.deleteByUserId(userId);
        labelRepository.deleteByUserId(userId);
    }
}

File: frontend/src/app/core/services/label.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { environment } from '../../../environments/environment';

export interface Label {
  id: number;
  userId: string;
  name: string;
  color: string;
  createdAt: string;
  updatedAt: string;
}

export interface MessageLabel {
  id: number;
  userId: string;
  messageUid: string;
  folder: string;
  label: Label;
  createdAt: string;
}

export interface LabelResponse {
  labels: Label[];
  total: number;
}

export interface MessageLabelResponse {
  messageLabels: MessageLabel[];
  total: number;
}

export interface LabelCreateRequest {
  name: string;
  color: string;
}

export interface LabelUpdateRequest {
  name: string;
  color: string;
}

export interface BatchLabelRequest {
  messageUids: string[];
  folder: string;
  labelIds: number[];
}

@Injectable({
  providedIn: 'root'
})
export class LabelService {

  constructor(private http: HttpClient) {}

  /**
   * Get all labels for the current user
   */
  getLabels(): Observable<LabelResponse> {
    return this.http.get<LabelResponse>(`${environment.apiUrl}/labels`)
      .pipe(
        catchError(error => {
          console.error('Error fetching labels:', error);
          return throwError(() => error);
        })
      );
  }

  /**
   * Get a specific label by ID
   */
  getLabel(id: number): Observable<{ label: Label }> {
    return this.http.get<{ label: Label }>(`${environment.apiUrl}/labels/${id}`)
      .pipe(
        catchError(error => {
          console.error('Error fetching label:', error);
          return throwError(() => error);
        })
      );
  }

  /**
   * Create a new label
   */
  createLabel(request: LabelCreateRequest): Observable<{ message: string; label: Label }> {
    const formData = new FormData();
    formData.append('name', request.name);
    formData.append('color', request.color);

    return this.http.post<{ message: string; label: Label }>(`${environment.apiUrl}/labels`, formData)
      .pipe(
        catchError(error => {
          console.error('Error creating label:', error);
          return throwError(() => error);
        })
      );
  }

  /**
   * Update an existing label
   */
  updateLabel(id: number, request: LabelUpdateRequest): Observable<{ message: string; label: Label }> {
    const formData = new FormData();
    formData.append('name', request.name);
    formData.append('color', request.color);

    return this.http.put<{ message: string; label: Label }>(`${environment.apiUrl}/labels/${id}`, formData)
      .pipe(
        catchError(error => {
          console.error('Error updating label:', error);
          return throwError(() => error);
        })
      );
  }

  /**
   * Delete a label
   */
  deleteLabel(id: number): Observable<{ message: string }> {
    return this.http.delete<{ message: string }>(`${environment.apiUrl}/labels/${id}`)
      .pipe(
        catchError(error => {
          console.error('Error deleting label:', error);
          return throwError(() => error);
        })
      );
  }

  /**
   * Get all messages with a specific label
   */
  getMessagesWithLabel(labelId: number): Observable<MessageLabelResponse> {
    return this.http.get<MessageLabelResponse>(`${environment.apiUrl}/labels/${labelId}/messages`)
      .pipe(
        catchError(error => {
          console.error('Error fetching messages with label:', error);
          return throwError(() => error);
        })
      );
  }

  /**
   * Apply a label to a message
   */
  applyLabelToMessage(labelId: number, messageUid: string, folder: string): Observable<{ message: string; messageLabel: MessageLabel }> {
    const formData = new FormData();
    formData.append('messageUid', messageUid);
    formData.append('folder', folder);

    return this.http.post<{ message: string; messageLabel: MessageLabel }>(`${environment.apiUrl}/labels/${labelId}/messages`, formData)
      .pipe(
        catchError(error => {
          console.error('Error applying label to message:', error);
          return throwError(() => error);
        })
      );
  }

  /**
   * Remove a label from a message
   */
  removeLabelFromMessage(labelId: number, messageUid: string, folder: string): Observable<{ message: string }> {
    const params = new HttpParams()
      .set('messageUid', messageUid)
      .set('folder', folder);

    return this.http.delete<{ message: string }>(`${environment.apiUrl}/labels/${labelId}/messages`, { params })
      .pipe(
        catchError(error => {
          console.error('Error removing label from message:', error);
          return throwError(() => error);
        })
      );
  }

  /**
   * Get all labels applied to a specific message
   */
  getMessageLabels(messageUid: string, folder: string): Observable<MessageLabelResponse> {
    const params = new HttpParams().set('folder', folder);

    return this.http.get<MessageLabelResponse>(`${environment.apiUrl}/labels/messages/${messageUid}`, { params })
      .pipe(
        catchError(error => {
          console.error('Error fetching message labels:', error);
          return throwError(() => error);
        })
      );
  }

  /**
   * Apply labels to multiple messages in batch
   */
  applyLabelsToMessages(request: BatchLabelRequest): Observable<{ message: string; processedMessages: number; appliedLabels: number }> {
    const formData = new FormData();
    request.messageUids.forEach(uid => formData.append('messageUids', uid));
    formData.append('folder', request.folder);
    request.labelIds.forEach(id => formData.append('labelIds', id.toString()));

    return this.http.post<{ message: string; processedMessages: number; appliedLabels: number }>(`${environment.apiUrl}/labels/batch/apply`, formData)
      .pipe(
        catchError(error => {
          console.error('Error applying labels to messages:', error);
          return throwError(() => error);
        })
      );
  }

  /**
   * Remove labels from multiple messages in batch
   */
  removeLabelsFromMessages(request: BatchLabelRequest): Observable<{ message: string; processedMessages: number; removedLabels: number }> {
    const formData = new FormData();
    request.messageUids.forEach(uid => formData.append('messageUids', uid));
    formData.append('folder', request.folder);
    request.labelIds.forEach(id => formData.append('labelIds', id.toString()));

    return this.http.post<{ message: string; processedMessages: number; removedLabels: number }>(`${environment.apiUrl}/labels/batch/remove`, formData)
      .pipe(
        catchError(error => {
          console.error('Error removing labels from messages:', error);
          return throwError(() => error);
        })
      );
  }

  /**
   * Get message UIDs for a specific label and folder
   */
  getMessageUidsWithLabel(labelId: number, folder: string): Observable<{ messageUids: string[]; total: number }> {
    const params = new HttpParams().set('folder', folder);

    return this.http.get<{ messageUids: string[]; total: number }>(`${environment.apiUrl}/labels/${labelId}/messages/uids`, { params })
      .pipe(
        catchError(error => {
          console.error('Error fetching message UIDs with label:', error);
          return throwError(() => error);
        })
      );
  }

  /**
   * Get usage count for a specific label
   */
  getLabelUsageCount(labelId: number): Observable<{ labelId: number; usageCount: number }> {
    return this.http.get<{ labelId: number; usageCount: number }>(`${environment.apiUrl}/labels/${labelId}/count`)
      .pipe(
        catchError(error => {
          console.error('Error fetching label usage count:', error);
          return throwError(() => error);
        })
      );
  }

  /**
   * Apply a single label to multiple messages
   */
  applyLabel(messageUids: string[], labelId: number, folder: string): Observable<{ message: string; processedMessages: number; appliedLabels: number }> {
    const request: BatchLabelRequest = {
      messageUids,
      folder,
      labelIds: [labelId]
    };
    return this.applyLabelsToMessages(request);
  }

  /**
   * Remove a single label from multiple messages
   */
  removeLabel(messageUids: string[], labelId: number, folder: string): Observable<{ message: string; processedMessages: number; removedLabels: number }> {
    const request: BatchLabelRequest = {
      messageUids,
      folder,
      labelIds: [labelId]
    };
    return this.removeLabelsFromMessages(request);
  }
}